pkg/agent/agent.go

```diff
@@ -0,0 +1,159 @@
+// Package agent implements functions for injecting faults in a target
+package agent
+
+import (
+	"context"
+	"fmt"
+	"syscall"
+	"time"
+
+	"github.com/grafana/xk6-disruptor/pkg/agent/protocol"
+	"github.com/grafana/xk6-disruptor/pkg/agent/protocol/grpc"
+	"github.com/grafana/xk6-disruptor/pkg/agent/protocol/http"
+	"github.com/grafana/xk6-disruptor/pkg/runtime"
+)
+
+// Config maintains the configuration for the execution of the agent
+type Config struct {
+	Profiler *runtime.ProfilerConfig
+}
+
+// Agent maintains the state required for executing an agent command
+type Agent struct {
+	env    runtime.Environment
+	config *Config
+}
+
+// BuildAgent builds a instance of an agent
+func BuildAgent(env runtime.Environment, config *Config) *Agent {
+	return &Agent{
+		env:    env,
+		config: config,
+	}
+}
+
+// do executes a command in the Agent
+func (r *Agent) do(ctx context.Context, action func(context.Context) error) error {
+	sc := r.env.Signal().Notify(syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)
+	defer func() {
+		r.env.Signal().Reset()
+	}()
+
+	acquired, err := r.env.Lock().Acquire()
+	if err != nil {
+		return fmt.Errorf("could not acquire process lock: %w", err)
+	}
+	if !acquired {
+		return fmt.Errorf("another instance of the agent is already running")
+	}
+
+	defer func() {
+		_ = r.env.Lock().Release()
+	}()
+
+	// start profiler
+	profiler, err := r.env.Profiler().Start(*r.config.Profiler)
+	if err != nil {
+		return fmt.Errorf("could not create profiler %w", err)
+	}
+
+	// ensure the profiler is closed even if there's an error executing the command
+	defer func() {
+		_ = profiler.Close()
+	}()
+
+	// set context for command
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// execute action goroutine to prevent blocking
+	cc := make(chan error)
+	go func() {
+		cc <- action(ctx)
+	}()
+
+	// wait for command completion or cancellation
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-cc:
+		return err
+	case s := <-sc:
+		return fmt.Errorf("received signal %q", s)
+	}
+}
+
+// ApplyHTTPDisruption applies a disruption to the gRPC requests serviced by the target
+func (r *Agent) ApplyHTTPDisruption(
+	ctx context.Context,
+	proxyConfig http.ProxyConfig,
+	disruption http.Disruption,
+	config protocol.DisruptorConfig,
+	transparent bool,
+	duration time.Duration,
+) error {
+	proxy, err := http.NewProxy(proxyConfig, disruption)
+	if err != nil {
+		return err
+	}
+
+	return r.applyDisruption(
+		ctx,
+		proxy,
+		config,
+		transparent,
+		duration,
+	)
+}
+
+// ApplyGrpcDisruption applies a disruption to the gRPC requests serviced by the target
+func (r *Agent) ApplyGrpcDisruption(
+	ctx context.Context,
+	proxyConfig grpc.ProxyConfig,
+	disruption grpc.Disruption,
+	config protocol.DisruptorConfig,
+	transparent bool,
+	duration time.Duration,
+) error {
```

Comment:
This large amount of arguments also makes me think this wants to be split.

I would suggest something like the following:

```go
// Run function for a command:
func (c *RootCommand) HTTPCmdRun() {
  // Create a protocol-specific proxy.
  proxy := http.NewProxy(c.proxyConfig, c.disruption)
  // Create a disruptor that will use the given proxy interface.
  disruptor := protocol.Disruptor(c.Env.Executor(), proxy)
  // Build an agent that will run the disruptor.
  agent := agent.BuildAgent(c.Env, c.agentConfig)

  return agent.Run(disruptor)
}
```